### PR TITLE
Fixes # 76094 - Check the content of WHAT for Makefile's verify target

### PR DESCRIPTION
**Pull request by YoubingLi**
**What type of PR is this?**

 /kind bug

**What this PR does / why we need it**:

If the target xx specified in WHAT is not verified, make verify WHAT="xx yy" should fail

**Which issue(s) this PR fixes**:

Fixes # 76094

**Special notes for your reviewer**:

/assign @liggitt
/assign @BenTheElder 

**Does this PR introduce a user-facing change?**:

no

```release-note
NONE
```

### CODE REVIEW

**Comment by YoubingLi** [comment]
#76094

File: hack/make-rules/verify.sh
Comment (SataQiu) [inline hack/make-rules/verify.sh]: Thanks @YoubingLi 
added some minor comments:

```
In expr_test.sh line 116:
    index=`expr $index + 1`
          ^-- SC2006: Use $(..) instead of legacy `..`.
           ^-- SC2003: expr is antiquated. Consider rewriting this using $((..)), ${} or [[ ]].
```

Comment (YoubingLi) [reply hack/make-rules/verify.sh]: Sure.

    Following code will be used.
                  index=$(($index + 1))

File: hack/make-rules/verify.sh
Comment (liggitt) [inline hack/make-rules/verify.sh]: can we run this first to fail fast if a bad target is specified?

**Comment by BenTheElder** [comment]
@YoubingLi I fixed the release note FYI, it needs to be `NONE` specifically for no release note (we parse this) :sweat_smile:

**Comment by BenTheElder** [comment]
skimmed this, what does the failure output look like?

**Comment by YoubingLi** [comment]
> skimmed this, what does the failure output look like?

It is similar as case failure

 make verify WHAT="ggg bb"
========================
FAILED TESTS
========================
ggg
bb
make: *** [verify] Error 1

File: hack/make-rules/verify.sh
Comment (YoubingLi) [reply hack/make-rules/verify.sh]: The reason is  the functionality of "is-explicitly-chosen" called by run-checks  can be leveraged to resolve this issue.

Currently we retrieve target list basing on the user's input paramter of run-checks.
If we want to run it before "run-checks", we need to ensure all verification test scripts are saved under "${KUBE_ROOT}/hack" directory forever.

Basing on the current fix, the end-user could get all failure information about both invalid and failure targets. For example 

     WHAT="xx godeps imports" 
Where
    xx:          the invalid target
    godeps: test would fail

The output  is

========================
FAILED TESTS
========================
godeps
xx

File: hack/make-rules/verify.sh
Comment (cblecker) [inline hack/make-rules/verify.sh]: Split string variable more robustly.

```suggestion
TARGET_LIST=()
IFS=" " read -r -a TARGET_LIST <<< "${WHAT}"
```

**Comment by YoubingLi** [comment]
> /test pull-kubernetes-godeps

@liggitt 
    staging-godeps test failure

    Seems like it relates to your current project. 
    Can you give me a clue to resolve it?

File: hack/make-rules/verify.sh
Comment (cblecker) [reply hack/make-rules/verify.sh]: It still won't fail fast, because we don't print/exit until below this. The verify job isn't currently designed to fail fast.

File: hack/make-rules/verify.sh
Comment (cblecker) [inline hack/make-rules/verify.sh]: Instead of looping through two distinct objects, just loop through the TARGET_LIST array

```suggestion
  for e in "${TARGET_LIST[@]}"; do
```

File: hack/make-rules/verify.sh
Comment (cblecker) [inline hack/make-rules/verify.sh]: $/${} is unnecessary on arithmetic variables.

```suggestion
    index=$((index + 1))
```

File: hack/make-rules/verify.sh
Comment (cblecker) [inline hack/make-rules/verify.sh]: Can we name this something like `missing-target-checks` instead so it's clear what it's doing. A comment wouldn't be a bad idea either.

**Review by cblecker — CHANGES_REQUESTED** [review]
This looks really great! Thanks for doing this. A few bash nits to fix up.

**Comment by cblecker** [comment]
@YoubingLi The failure in the `pull-kubernetes-godeps` job is legitimate.. it's your changes picking up a verify test that the job is attempting, but doesn't exist anymore. Should be fixed when https://github.com/kubernetes/test-infra/pull/12104 merges.

File: hack/make-rules/verify.sh
Comment (YoubingLi) [reply hack/make-rules/verify.sh]: Good. I will revise it.

File: hack/make-rules/verify.sh
Comment (YoubingLi) [reply hack/make-rules/verify.sh]: I will address it in the latest diff.

File: hack/make-rules/verify.sh
Comment (YoubingLi) [reply hack/make-rules/verify.sh]: I will address it.

**Comment by YoubingLi** [comment]
When WHAT is not specified, there is code error 

        IFS=" " read -r -a TARGET_LIST <<< "${WHAT}"

It should be 
        IFS=" " read -r -a TARGET_LIST <<< "${WHAT:-}"

Meanwhile, the number of element in TARGET_LIST should be check in missing-target-checks

**Comment by YoubingLi** [comment]
> /test pull-kubernetes-godeps

@cblecker 

   verify-godep-licenses.sh and verify-godeps.sh are successed.
   The test complains "staging-godeps"

    kubernetes/test-infra#12104 should fix it, right?

        FAILED TESTS
        staging-godeps
        Makefile:128: recipe for target 'verify' failed
        make: *** [verify] Error 1

**Comment by cblecker** [comment]
/lgtm
/approve
/hold

Yes, that PR should fix that job. Going to place hold on this until that's merged and we can check.

**Review by cblecker — APPROVED** [review]
/lgtm
/approve

Thanks for this, @YoubingLi!

**Pull request merged**
